lib/seahorn/FiniteMapTransf.cc

```diff
@@ -0,0 +1,351 @@
+  // `ks1` and `ks2` are needed.
+  Expr mkEqCore(Expr m1, Expr ks1, Expr m2, Expr ks2, ExprVector &keys, Expr vTy,
+                ExprFactory &efac, ExprSet &evars) {
+
+    ExprVector conj;
+
+    bool is_var_m1 = bind::isFiniteMapConst(m1);
+    bool is_var_m2 = bind::isFiniteMapConst(m2);
+
+    Expr e_m1, e_m2;
+
+    for (auto k : keys) {
+      if (is_var_m1) {
+        e_m1 = mkVarGet(m1, k, vTy);
+        evars.insert(e_m1);
+      } else
+        e_m1 = finite_map::mkGetVal(m1, ks1, k);
+
+      if (is_var_m2) {
+        e_m2 = mkVarGet(m2, k, vTy);
+        evars.insert(e_m2);
+      } else
+        e_m2 = finite_map::mkGetVal(m2, ks2, k);
+      conj.push_back(mk<EQ>(e_m1, e_m2));
+    }
+    return mknary<AND>(conj);
+  }
+
+  Expr mkInitFMapCore(Expr map, ExprMap &type_lambda, ExprMap &expr_type,
+                      ExprFactory &efac) {
+
+    // defmap(defk(keys), default(valTy)))
+    //      or
+    // defmap(defk(keys), defv(values)))
+
+    // build keys
+    Expr defk = map->left();
+    assert(isOpX<CONST_FINITE_MAP_KEYS>(defk));
+    ExprVector keys(defk->args_begin(), defk->args_end());
+
+    Expr vTy, valuesE = map->right();
+    if (isOpX<FINITE_MAP_VAL_DEFAULT>(valuesE)) { // non init values
+      vTy = valuesE->left();                      // type of the values
+    } else {
+      assert(isOpX<CONST_FINITE_MAP_VALUES>(valuesE)); // initialized values
+      vTy = bind::typeOf(*valuesE->args_begin());
+      // already expanded, can they be of unknown type?
+    }
+
+    Expr fmTy = sort::finiteMapTy(vTy, keys);
+    expr_type[map] = fmTy;
+    Expr lmdKeys = finite_map::mkKeys(keys, efac);
+    type_lambda[fmTy] = lmdKeys;
+
+    return map;
+  }
+
+  Expr mkFMapPrimitiveArgCore(Expr map, ExprMap &type_lambda,
+                              ExprMap &expr_type, ExprFactory &efac) {
+
+    if (isOpX<CONST_FINITE_MAP>(map)){
+      Expr defk = map->left();
+      assert(isOpX<CONST_FINITE_MAP_KEYS>(defk));
+      Expr lmdKeys = type_lambda[map];
+      Expr fmTy = expr_type[map];
+      Expr res, valuesE = map->right();
+      Expr vTy = finite_map::valTy(fmTy);
+      if (isOpX<FINITE_MAP_VAL_DEFAULT>(valuesE)) { // non init values
+        return finite_map::mkEmptyMap(vTy, efac);
+      } else {
+        assert(isOpX<CONST_FINITE_MAP_VALUES>(valuesE)); // initialized values
+        ExprVector values(valuesE->args_begin(), valuesE->args_end());
+        ExprVector keys(defk->args_begin(), defk->args_end());
+        return finite_map::mkInitializedMap(keys, vTy, values, efac, lmdKeys);
+      }
+    }
+    else // already transformed map: 0 or ite expr
+      return map;
+  }
+
+  // \brief expands a map into separate scalar variables
+  void mkVarsMap(Expr map, Expr lmdks, ExprVector &keys, Expr vTy,
+                 ExprVector &new_vars, ExprVector &extra_unifs,
+                 ExprFactory &efac, ExprSet &evars) {
+
+    Expr v, v_get;
+    ExprVector map_values;
+
```

Comment:
can reserve space for `new_vars` and `map_values` before the loop to avoid resizing

Reply:
Done for `map_values`. For `new_vars`, it may already contain things because, as it accumulates all the new variables (from other arguments that are maps). I can create a new `ExprVector` and then push its content to `new_vars`, is that more efficient than what there is now?